app/code/Magento/Catalog/Block/Adminhtml/Product/Edit/Action/UrlBuilder.php

```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+/**
+ * Adminhtml catalog product action customer view
+ *
+ * @author      Marcin Dykas <mdykas@divante.pl>
+ */
+namespace Magento\Catalog\Block\Adminhtml\Product\Edit\Action;
+
+use Magento\Store\Api\StoreResolverInterface;
+
+class UrlBuilder
+{
+    /**
+     * @var \Magento\Framework\UrlInterface
+     */
+    protected $frontendUrlBuilder;
+
+    /**
+     * @param \Magento\Framework\UrlInterface $frontendUrlBuilder
+     */
+    public function __construct(\Magento\Framework\UrlInterface $frontendUrlBuilder)
+    {
+        $this->frontendUrlBuilder = $frontendUrlBuilder;
+    }
+
+    /**
+     * Get action url
+     *
+     * @param string $routePath
+     * @param \Magento\Catalog\Model\Product $product
+     * @param string $scope
+     * @param string $store
+     * @return string
+     */
+    public function getUrl($routePath, $product, $scope, $store)
```

Comment:
If `$product` needs the `getId` method then I would suggest adding a type hint here to enforce the type of object passed in. In this case I would suggest the `\Magento\Catalog\Api\Data\ProductInterface` as this will enforce the `getId` method implementation.